draft-wang-ppm-differential-privacy.md

```diff
@@ -255,12 +255,58 @@ with the correct noise type.
 > TODO: Chris P: we will mention Prio3SumVec because that's what we use to describe aggregator DP with amplification
 
 
-# Differential Privacy Mechanism {#dp mechanism}
+# Differential Privacy Mechanism {#dp-mechanism}
 
 This section should define a framework to apply different dp noise type to a
 client measurement and/or an aggregation.
 
-> TODO: Junye to spell out the interface of dp mechanisms, "Noise Mechanisms to Accomplish DP"
+Clients and Aggregators can implement various noise mechanisms to accomplish
+the DP guarantees described in {{budget}}. The parameters of the noise
+mechanisms should be computed based on the target DP guarantee, and can be
+passed to the initializers of the noise mechanisms.
+
+We define three methods for the noise mechanism interface `DpMechanism`:
+
+* `add_noise` that adds noise to a piece of data. This can be used by a Client
+  to add noise to its measurement.
+
+* `sample_noise` that allows `DpMechanism` to sample noise directly. This can
+  be used by an Aggregator to sample Aggregator DP noise, which then gets merged
+  with the Aggregator's aggregate share.
+
+* `debias` that allows `DpMechanism` to debias the data due to the bias
+  generated by the noise. Note not all noise will need this functionality.
+  We currently think some local randomizers will need this functionality.
+  For example, Appendix A and Appendix C of {{MJTBCCDFGHJKLLMPPPPPRSWZ22}}
+  describe two types of randomized response local randomizers that will
+  need the debiasing functionality.
+
+~~~
+class DpMechanism:
+    DataType
+    DebiasedDataType
+
+    def add_noise(self, data: DataType) -> DataType:
+        """Add noise to a piece of data. """
+        pass
+
+    def sample_noise(self, num_reps: int, dimension: int) -> DataType:
```

Comment:
IIUC this would be called by add_noise, i.e., is more of a DP primitive than a DP mechanism. Can we rename it?

Reply:
I think this gives caller flexibility to encode the returned noise differently, e.g. when sampling noise in the VDAF setting, one will sample the noise with this method, and then encode it to into field elements. To take DiscreteGaussian as an example, the returned data is an array of signed integers, and there is another step that converts the negative values to the back end of field modulus (e.g. -1 to `q-1`, where `q` is field modulus).

Reply:
Think of discrete guassian as a primitive that we use to implement `add_noise_to_agg_share()` or `add_noise_to_measurement()`. We would invoke that primitive in both of these methods, then do any postprocessing that's needed to fit with the VDAF. Does that make sense?

Reply:
Sounds good, moved those logic into `DpMechanism`. `DpPrimitive`'s `sample_noise` will only deal with generating noise.